src/main/scala/com/github/pathikrit/metarest/Resource.scala

```diff
@@ -1,74 +1,55 @@
-          case (q"new $annotation", f) if Set("get", "post", "put") contains annotation.toString => annotation.toString -> f.duplicate
-        }
-      }
-      newFields.groupBy(_._1) map { case (annotation, values) =>
-        val (className, classFields) = (TypeName(annotation.capitalize), values.map(_._2))
-        q"case class $className(..$classFields)"
-      }
+@compileTimeOnly("@metarest.Resource not expanded")
+class Resource extends StaticAnnotation {
+  inline def apply(defn: Any): Any = meta {
+    val (cls: Defn.Class, companion: Defn.Object) = defn match {
+      case Term.Block(Seq(cls: Defn.Class, companion: Defn.Object)) => (cls, companion)
+      case cls: Defn.Class => (cls, q"object ${Term.Name(cls.name.value)} {}")
+      case _ => abort("@metarest.Resource must annotate a class")
     }
 
-    def modifiedDeclaration(classDecl: ClassDef, compDeclOpt: Option[ModuleDef] = None) = {
-      val (className, fields) = getClassAndFields(classDecl)
-
-      val compDecl = compDeclOpt map { compDecl =>
-        val q"object $obj extends ..$bases { ..$body }" = compDecl
-        q"""
-          object $obj extends ..$bases {
-            ..$body
-            ..${generateModels(fields)}
+    val paramsWithAnnotation = for {
+      Term.Param(mods, name, decltype, default) <- cls.ctor.paramss.flatten
+      seenMods = mutable.Set.empty[String]
+      modifier <- mods if seenMods.add(modifier.toString)
+      (tpe, defArg) <- modifier match {
+        case mod"@get" | mod"@put" | mod"@post" => Some(decltype -> default)
+        case mod"@patch" =>
+          val optDeclType = decltype.collect({case tpe: Type => targ"Option[$tpe]"})
+          val defaultArg = default match {
+            case Some(term) => q"Some($term)"
+            case None => q"None"
           }
-        """
-      } getOrElse {
-        q"""
-          object ${className.toTermName} {
-            ..${generateModels(fields)}
-          }
-         """
+          Some(optDeclType -> Some(defaultArg))
+        case _ => None
       }
-
-      c.Expr( q"""
-        $classDecl
-        $compDecl
-      """)
-    }
-
-    withCompileError("must annotate a class") {
-      annottees.map(_.tree) match {
-        case (classDecl: ClassDef) :: Nil => modifiedDeclaration(classDecl)
-        case (classDecl: ClassDef) :: (compDecl: ModuleDef) :: Nil => modifiedDeclaration(classDecl, Some(compDecl))
-      }
-    }
+    } yield modifier -> Term.Param(Nil, name, tpe, defArg)
+
+    val models = paramsWithAnnotation
+      .groupBy(_._1.toString)
+      .map({case (verb, pairs) =>
+        val className = Type.Name(verb.stripPrefix("@").capitalize)
+        val classParams = pairs.map(_._2)
+        q"case class $className[..${cls.tparams}] (..$classParams)"
+      })
+
+    val newCompanion = companion.copy(
+      templ = companion.templ.copy(stats = Some(
+        companion.templ.stats.getOrElse(Nil) ++ models
+      ))
+    )
```

Comment:
This is where https://github.com/scalameta/scalameta/issues/741 should be highly relevant. /cc @olafurpg @DavidDudson